Add variations

marcrowo opened the pull request:
~~**NOTE**: This branch is a result of merging #31 `make-it-a-shield-study` and #32 `limit-times-shown-via-pref`.~~

**NOTE**: Tests may fail as there are changes in functionality.

Variations are centered on adding aggressiveness of adding button to toolbar.
1. If the share button is not in the toolbar, then don’t do anything upon URL bar copy.
2. If the share button is not in the toolbar, then upon URL bar copy ask to move the share button study to the toolbar (with a panel on burger menu or page action on URL bar).
3. If the share button is not in the toolbar, then move to the toolbar without asking.
It’s important to keep track of which case and what we did so we can undo this change after the study is over (?).

marcrowo committed:
Added first implementation of aggressive adding button to toolbar.

marcrowo committed:
Fixed typo and linting error.

marcrowo committed:
Added doorhangerAskToAdd variation.

marcrowo committed:
Add telemetry pings for asking to add and adding without asking.

marcrowo committed:
WIP: adding variations + testing.

marcrowo committed:
Added HTML/CSS for asking panel.

marcrowo committed:
Fleshed out highlight tests, some tests for other treatments.

marcrowo committed:
Moving each treatment test into a new describe() with a separate Firefox process.

marcrowo committed:
Added DoorHangerAskToAdd tests.

marcrowo committed:
Fixed opening ask panel if the share button has already been added.

marcrowo committed:
Added tests for addToToolbar treatment.

marcrowo committed:
Added MAX_TIMES_TO_SHOW check to Basic Functional Tests.

marcrowo committed:
Added uninstall tests to Basic Functional Tests.

marcrowo commented:
### New test list (taken from [build 45](https://circleci.com/gh/marcrowo/share-button-study/45?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link))

**Basic Functional Tests**
    ✓ should have a URL bar 
    ✓ should have a share button
    ✓ should have copy paste working
    ✓ should only trigger MAX_TIMES_TO_SHOW = 5 times
    _Addon uninstall tests_
      ✓ should no longer trigger animation once uninstalled
      ✓ should no longer trigger popup once uninstalled 
      ✓ should no longer trigger ask panel once uninstalled
      ✓ should not add the button to the toolbar once uninstalled

  **Highlight Treatment Tests**
    ✓ animation should trigger on regular page 
    ✓ animation should not trigger on disabled page
    ✓ animation should not trigger if the share button is not added to toolbar 
    ✓ should not trigger treatments if the share button is in the overflow menu 
    ✓ should send highlight and copy telemetry pings 

  **DoorhangerDoNothing Treatment Tests**
    ✓ popup should trigger on regular page 
    ✓ popup should not trigger on disabled page
    ✓ popup should not trigger if the share button is not added to toolbar 
    ✓ should not trigger treatments if the share button is in the overflow menu
    ✓ should send doorhanger and copy telemetry pings

  **DoorhangerAskToAdd Treatment Tests**
    ✓ should open an ask panel on a regular page without the share button 
    ✓ should open a standard panel on a regular page with the share button
    ✓ should not open an ask panel on a regular page with the share button
    ✓ should not open an ask panel on a regular page if the share button is in the overflow menu
    ✓ should not open an ask panel on a disabled page
    ✓ should add the button to the toolbar upon clicking on ask panel
    ✓ should send ask-to-add and copy telemetry pings

  **DoorhangerAddToToolbar Treatment Tests**
    ✓ should add the button to the toolbar upon copy paste on regular page
    ✓ popup should trigger on regular page
    ✓ should send add-to-toolbar and copy telemetry pings

  **New Window Add-on Functional Tests**
    ✓ animation should trigger on regular page

marcrowo commented:
**FIXME**: The tests still crash in some cases, ie. `Components.utils is not defined`.

marcrowo committed:
Refactored browser navigation into utils.gotoURL().

marcrowo committed:
Fixed linting issue.

mythmon commented on extension/ask.html:
```diff
@@ -0,0 +1,22 @@
+            </div>
+            <div class="body">
+                Want to share links quickly and easily? <br>
+                Add the share button to your toolbar by clicking anywhere in this message!
```
Instead of the `<br>` tag, it would probably be better to use something more semantically meaningful, like a pair of `<p>` tags:

```html
<div class="body">
  <p>Want to share links quickly and easily?</p>
  <p>Add the share button to your toolbar by clicking anywhere in this message!</p>
</div>
```

mythmon commented on extension/bootstrap.js:
```diff
@@ -19,59 +20,136 @@ const REASONS = {
+      browserWindow.window.document.getElementById("mainPopupSet").appendChild(panel);
+    }
+    const burgerMenu = browserWindow.window.document.getElementById("PanelUI-menu-button");
+    // TODO What if there is no burger menu?
```
I think it is safe to always assume there is a burger menu. I don't think there is anyway to remove it.

mythmon commented on extension/bootstrap.js:
```diff
@@ -19,59 +20,136 @@ const REASONS = {
+    // was initialized before the button was added
+    doorhangerDoNothingTreatment(browserWindow, browserWindow.shareButton);
+  } else if (shareButtonIsUseable(shareButton)) {
+    doorhangerDoNothingTreatment(browserWindow, browserWindow.shareButton);
```
Since you call `doorhangerDoNothingTreatment` either way, I'd move it out of the else branch, and call it unconditionally at the end of the function. It's a bit clearer that way.

mythmon commented on extension/bootstrap.js:
```diff
@@ -19,59 +20,136 @@ const REASONS = {
+  if (Preferences.isSet(TREATMENT_OVERRIDE_PREF)) {
+    variation = { name: Preferences.get(TREATMENT_OVERRIDE_PREF, variation),
+      weight: 1 };
+    Preferences.reset(TREATMENT_OVERRIDE_PREF); // reset to default branch value to clear
```
Why do you want to clear the pref?

marcrowo replied:
This is to "delete" the preference so that it does not remain after the extension is uninstalled.

I think that I was thinking that the variation would only be used once upon install, but the addon actually calls `chooseVariation()` in the bootstrap `startup()`, which means that it will be called upon install and each time Firefox is started. This has not been an issue so far since I only use the `TREATMENT_OVERRIDE_PREF` in tests, for which I am usually not restarting Firefox or I set the pref before each test.

Do you think it would make more sense to move this to the `shutdown()` function, like where I reset the `COUNTER_PREF`? (bootstrap.js:406)

mythmon replied:
I think that this shouldn't actually get cleared. It is ok to clear the counter pref, since that is really internal state. But this is user choice (well, tester choice) that I wouldn't expect to get poofed just because the add-on got uninstalled. It also isn't a matter of cleaning up after the extension, since this pref will only be set manually.

mythmon commented on extension/bootstrap.js:
```diff
@@ -91,25 +169,12 @@ class CopyController {
+      // check to see if we should call a treatment at all
+      const numberOfTimeShown = Preferences.get(COUNTER_PREF, 0);
+      if (numberOfTimeShown < MAX_TIMES_TO_SHOW &&
+          Object.keys(TREATMENTS).includes(this.treatment)) {
```
When do you expect this to be false?

marcrowo replied:
This should be false if either the treatment has been shown >= `MAX_TIMES_TO_SHOW` or if the treatment is not valid.

mythmon replied:
I was specifically referring to the second part, the `Object.keys(TREATMENTS).include(this.treatment)`. What would cause `this.treatment` to not be a valid treatment name? I don't see any way that could happen, and either way you probably shouldn't be checking that sort of data at this point. It should be prevented from getting into this weird state in the first place.

marcrowo replied:
This could happen because the treatment pref is set to an invalid treatment or the config file has invalid treatment names, where the valid treatment names are given by the `TREATMENTS` object in `bootstrap.js`. There is no other check to make sure the treatments that are set in the pref or config file are valid.

However, I think that you are right that this check should not occur right before the treatment function is being called and instead be called somewhere more appropriate such as after we choose the variation.

marcrowo replied:
Per discussion in person, we decided that it is important to guard against user input (ie. the pref) but not so much against your own code (ie. config file).

So we decided to fall back to the config file if the pref is incorrect and throw an exception if the config file is incorrect.

mythmon commented on test/button_test.js:
```diff
@@ -94,110 +114,357 @@ describe("Add-on Functional Tests", function() {
-    await driver.get("about:blank");
-    driver.setContext(Context.CHROME);
+    await utils.addShareButton(driver);
+    await utils.gotoURL(driver, "about:blank");
```
This is a good utility. 👍

mythmon commented on test/utils.js:
```diff
@@ -206,9 +223,9 @@ module.exports.closePanel = async(driver) => {
 module.exports.getMostRecentPingsByType = async(driver, type) =>
-  driver.executeAsyncScript(async() => {
-    const typeArg = arguments[0];
-    const callback = arguments[arguments.length - 1];
```
If `...args` works, I think you should be able to put your parameters straight into the function declaration, like `driver.executeAsyncScript(async (typeArg, callback) => {`. Does that work?

marcrowo replied:
Yes, this works! I had tried to get this to work previously, but I think I had the order switched (ie. `async(callback, typeArg)`) so I decided to use `...args` instead.

marcrowo committed:
Updated variation names and changed pref reset logic.

mythmon reviewed (changes requested):
Agh, I started this review ages ago, and then forgot to submit it. Sorry about that! This might be a bit stale, so if something I said seems wrong, it might be!

marcrowo committed:
Change <br> to <p> in ask panel.

Also fixed to properly use ask.css and made p margins 0px.

marcrowo committed:
Stylistic improvement for doorhangerAddToToolbar.

Since we are calling the treatment either way, move it out of the else branch.

marcrowo committed:
Stylistic improvements to executeAsyncScript arg handling.

marcrowo committed:
Removed burger menu comment.

marcrowo committed:
Review fixes for variation choosing and variation pref.

- Check for invalid pref earlier in chooseVariation() instead of in the
  CopyController.
- Do not clear the user-set variation preference.
